File: introduction/pureimpure.hs

product' :: (Num a) => a -> a -> a
product' x y = x * y


getNumberFromUser :: IO(Int)
getNumberFromUser = do
  putStrLn("Enter an integer:")
  line <- getLine
  return (read line :: Int)

showResult :: String -> Int -> IO()
showResult operation result = 
  putStrLn ("The " ++ operation ++ " is: " ++ show result)

main :: IO()
main = do
  x <- getNumberFromUser
  y <- getNumberFromUser
  let result = product' x y 
  showResult "product" result

File: introduction/pureimpure.scala
trait Operations {
  def product(x: Int, y: Int) = x * y
}

trait Console {
  import scala.io.Source

  def getNumberFromUser(): Int = {
    println("Enter an integer:")
    Source.stdin.getLines.next.toInt
  }

  def showResult(operation: String, result: Int): Unit =
    println(s"The ${operation} is: ${result}")
}

object Productifier extends App with Operations with Console {
  
  val x = getNumberFromUser()
  val y = getNumberFromUser()
  val result = product(x, y)
  showResult("product", result)
}

